### DISCUSSION TITLE
Get Body if bodyParser is Disabled

### DISCUSSION THREAD

**Discussion by stepanushariara** [vercel/next.js#12517] [Help] [answered]
Hi guys,

How to get request body if the bodyParser is Disabled ?

**Comment by Timer** [comment]
Why did you disable the body parser?

If you disable Next.js' body parsing, you have to read the body yourself.
The request is a normal Node.js request object, so you can read it by following Node.js' docs:

https://nodejs.org/es/docs/guides/anatomy-of-an-http-transaction/#request-body

I'd highly suggest turning the body parser back on if you need the body. The only cases it should be disabled is when you're using a library that supports reading the body itself.

**Reply by hzik** [reply]
Hi Timer, it's due to security. All the requests comes with computed hash in the header and we need to take the body of such request and calculate hash using our secret. If those hashes match we are sure the request came from trusted source. So we need to access original body with original formatting to be able to compute hash correctly.

**Reply by Timer** [reply]
You can disable the body parser and read the instructions above. Next.js does not provide a helper to get the body as a Buffer.

This package may help:
https://www.npmjs.com/package/raw-body

**Comment by leerob** [comment]
```js
import type { NextApiRequest, NextApiResponse } from 'next';
import type { Readable } from 'node:stream';

export const config = {
  api: {
    bodyParser: false,
  },
};

async function buffer(readable: Readable) {
  const chunks = [];
  for await (const chunk of readable) {
    chunks.push(typeof chunk === 'string' ? Buffer.from(chunk) : chunk);
  }
  return Buffer.concat(chunks);
}

export default async function (req: NextApiRequest, res: NextApiResponse) {
  if (req.method === 'POST') {
    const buf = await buffer(req);
    const rawBody = buf.toString('utf8');

    res.json({ rawBody });
  } else {
    res.setHeader('Allow', 'POST');
    res.status(405).end('Method Not Allowed');
  }
}
```

[Source](https://vercel.com/support/articles/how-do-i-get-the-raw-body-of-a-serverless-function)

**Reply by IHIutch** [reply]
This seems to silently fail for me nor let me console.log anything and causes my response to timeout. Do I need to have a particular version of Next?

But if I use `buffer` from [micro](https://www.npmjs.com/package/micro), that works. 

Example: https://github.com/dijonmusters/build-a-saas-with-next-js-supabase-and-stripe/blob/master/26-configure-stripe-for-production-and-deploy-next-js-application-with-vercel/pages/api/stripe-hooks.js

**Reply by leerob** [reply]
I did a fresh `npx create-next-app@latest test-app --typescript` and pasted the code into `pages/api/hello.ts` and it worked to confirm.

**Reply by leerob** [reply]
```
curl -X POST -H "Content-Type: text/plain" --data "This is raw data" http://localhost:3000/api/hello
```

**Reply by IHIutch** [reply]
Good to know, thanks, I'll give that a whirl.

**Reply by lukeweston1234** [reply]
We were doing this for sometime, but recently we get build errors that this type of config is deprecated. I cannot find a replacement in the new documentation. Can anyone point me in the right direction?

**Comment by kiranvj** [comment]
I have `bodyParser : false`

and I used [formidable ](https://www.npmjs.com/package/formidable) to parse the request

  ```
const form = new formidable.IncomingForm();

  form.parse(req, async function (err, fields, files) {
    console.log('fields: ', fields);
  });
```

**Comment by eddiedotdev** [comment]
In my case, I was already converting the raw-body to validate a shopify header

```
const rawBody = await getRawBody(req);
const hmac = crypto
     .createHmac('sha256', process.env.SHOPIFY_SECRET as string)
     .update(rawBody)
     .digest('base64');
```

I was able to just parse and stringify it from the `rawBody` variable:

`const body = JSON.parse(rawBody.toString());`

**Reply by Jobin-S** [reply]
I am getting an error req is not stream

```
TypeError: argument stream must be a stream
    at getRawBody (webpack-internal:///(rsc)/./node_modules/raw-body/index.js:56:15)
```

**Reply by Jore** [reply]
Yeah same.

**Reply by Jore** [reply]
This might be something to do with using the new app router in nextjs.

I got it to work. Instead of using `raw-body` package  use ` const body = await req.text();`.

**Reply by imalmamun** [reply]
Thank you.  It helped me brother..

**Comment by leerob** [comment, marked as answer]
Updated answer: I'd recommend using Route Handlers, since you don't need to modify the configuration at all. To read the request body, you can use the standard Web APIs. You can use Route Handlers in the App Router even if the rest of your application is still on the Pages Router.

```js
export async function POST(request: Request) {
  const res = await request.text()
  return Response.json({ res })
}
```

https://nextjs.org/docs/app/building-your-application/routing/route-handlers

**Reply by tecoad** [reply]
@leerob Sometimes I get null body from these requests even the middleware api router is not matched in the middleware. Very though to debug, the only thing that clears it completely is removing the middleware.
